index.rst

```diff
@@ -1,29 +1,375 @@
+This in theory could be done via complicated rules in the ingress specifying how to match the URL and verb patterns of complex queries, but in practice that seems hard to maintain.
+Alternately, we could assume that all ``GET`` requests are cheap and all requests with other verbs are expensive, but unfortunately IVOA standards require some expensive queries be accesible via ``GET``.
+
+The current approach is the simplest and provides a general facility to impose basic rate-limits on anything, so we're going to start with it and see if it's adequate in practice.
+If not, we may need to move more quota checking into the separate services rather than in Gafaelfawr.
+
+Quota checking
+==============
+
+API
+---
+
+The ``/auth/api/v1/user-info`` route will be extended to add quota information.
+The response will look like this:
+
+.. code-block:: json
+
+  {
+      "username": "someuser",
+      "name": "Alice Example",
+      "email": "alice@example.com",
+      "uid": 4123,
+      "gid": 4123,
+      "groups": [
+          {
+              "name": "g_special_users",
+              "id": 123181
+          }
+      ],
+      "quota": {
+          "api": {
+              "datalinker": 500,
+              "hips": 2000,
+              "tap": 500,
+              "vo-cutouts": 100
+          },
+          "notebook": {
+              "cpu": 9,
+              "memory": "27Gi"
+          }
+      }
+  }
+
+The quota shown will be the calculated amount reflecting any additions from groups and any configured overrides.
+The sources of the quota components will not be shown.
+(We may eventually want to add a separate API to see the full quota breakdown of why a user has the quota that they do, but it's not part of the initial design.)
+
+Notebook Aspect
+---------------
+
+The Notebook Aspect lab controller (see :sqr:`066`) will use its delegated notebook token during menu creation and lab creation to retrieve the user's quota information.
+For the menu response, it will filter out any notebook sizes that exceed the user's quota.
+For the lab creation, it will add a Kubernetes ``ResourceQuota`` resource for the user's namespace that sets limits matching the user's quota.
+
+.. _rate-limiting:
+
+Rate limiting
+-------------
+
+Currently, a ``GafaelfawrIngress`` only configures the name of the protected service when it is requesting a delegated token (as ``config.delegate.internal.service``).
+This configuration will be moved up to ``config.service`` and correspond to a new ``service`` parameter to the ``/auth`` route, replacing ``delegate_to``.
+Delegation will then be controlled by ``delegate_scopes``.
+
+Rate limiting will then be done if and only if there is an API quota for a service whose name matches the ``service`` parameter.
+
+To support multiple Gafaelfawr instances without confusing impact on quotas from having quotas be tracked separately by different instances, the data for quota enforcement will be stored in Redis.
```

Comment:
This sentence is hard to parse.  "Quota enforcement data will be stored in Redis in order to support multiple Gafaelfawr instances from a single data quota source," or something?